[C10D] Support group ranks in P2POp and batch_isend_irecv

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #141054

Changes semantic of __repr__ of P2POp: s, d are now group ranks instead
of global ranks. I think this is OK since I also updated the field names
to make this obvious.

Also add mypy annotations

Partially addresses RFC 0042 (pytorch/rfcs#71)
See more details/motivation in #140460

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @d4l3k @c-p-i-o

wconstab committed:
[C10D] Support group ranks in P2POp and batch_isend_irecv

Changes semantic of __repr__ of P2POp: s, d are now group ranks instead
of global ranks. I think this is OK since I also updated the field names
to make this obvious.

Also add mypy annotations

Partially addresses RFC 0042 (pytorch/rfcs#71)
See more details/motivation in #140460

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Support group ranks in P2POp and batch_isend_irecv"


Changes semantic of __repr__ of P2POp: s, d are now group ranks instead
of global ranks. I think this is OK since I also updated the field names
to make this obvious.

Also add mypy annotations

Partially addresses RFC 0042 (pytorch/rfcs#71)
See more details/motivation in #140460

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 d4l3k c-p-i-o

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Support group ranks in P2POp and batch_isend_irecv"


Changes semantic of __repr__ of P2POp: s, d are now group ranks instead
of global ranks. I think this is OK since I also updated the field names
to make this obvious.

Also add mypy annotations

Partially addresses RFC 0042 (pytorch/rfcs#71)
See more details/motivation in #140460

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -469,57 +469,61 @@ class P2POp:
-        self.peer = peer
-        self.group = group
+        self.group = _group_or_default_group(group)
+        self.peer = _canonicalize_group_rank(
+            self.group, peer, group_peer, return_global=True
+        )
         self.tag = tag
+        self.group_peer = _canonicalize_group_rank(self.group, peer, group_peer)
```
nit: I saw a couple patterns like this. Would it be easier if we ask `_canonicalize_group_rank` to return both?

wconstab replied:
i thought about this, but then there are a lot of callsites where it only uses one.  then it would look like 

`_, group_dst = _canonicalize()`

or 

`dst, _ = _canonicalize()`

one downside there is its less obvious which side of the tuple means what. Wdyt?

kwen2501 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -2593,13 +2597,39 @@ def batch_isend_irecv(p2p_op_list):
         with _coalescing_manager(group, device, async_ops=True) as cm:
             for p2p_op in p2p_op_list:
-                p2p_op.op(p2p_op.tensor, p2p_op.peer, p2p_op.group, p2p_op.tag)
+                if p2p_op.op == isend:
+                    p2p_op.op(
+                        p2p_op.tensor,
+                        group=p2p_op.group,
+                        tag=p2p_op.tag,
+                        group_dst=p2p_op.group_peer,
+                    )
+                else:
+                    p2p_op.op(
+                        p2p_op.tensor,
+                        group=p2p_op.group,
+                        tag=p2p_op.tag,
+                        group_src=p2p_op.group_peer,
+                    )
```
nit:
```
if p2p_op.op == isend:
  peer_kwarg = {"group_dst": p2p_op.group_peer}
else:
  peer_kwarg = {"group_src": p2p_op.group_peer}

p2p_op.op(p2p_op.tensor, p2p_op.group, p2p_op.tag, **peer_kwarg)
```

kwen2501 reviewed (approved):
LGTM! Thanks!

wconstab committed:
Update on "[C10D] Support group ranks in P2POp and batch_isend_irecv"


Changes semantic of __repr__ of P2POp: s, d are now group ranks instead
of global ranks. I think this is OK since I also updated the field names
to make this obvious.

Also add mypy annotations

Partially addresses RFC 0042 (pytorch/rfcs#71)
See more details/motivation in #140460

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 d4l3k c-p-i-o

[ghstack-poisoned]

wconstab commented:
@pytorchbot merge -i
